pkg/engine/shield/add_test.go

```diff
@@ -0,0 +1,362 @@
+package shield
+
+import (
+	"testing"
+
+	"github.com/golang/mock/gomock"
+	"github.com/simimpact/srsim/pkg/engine/event"
+	"github.com/simimpact/srsim/pkg/engine/info"
+	"github.com/simimpact/srsim/pkg/engine/prop"
+	"github.com/simimpact/srsim/pkg/key"
+	"github.com/simimpact/srsim/pkg/model"
+	"github.com/stretchr/testify/assert"
+
+	"github.com/simimpact/srsim/tests/mock"
+)
+
+// Unit Tests for AddShield()
+func TestShieldHealthByPositiveValues(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	engine := mock.NewMockEngineWithEvents(mockCtrl)
+	attr := mock.NewMockAttribute(mockCtrl)
+	manager := New(engine.Events(), attr)
+
+	defer mockCtrl.Finish()
+
+	source := key.TargetID(1)
+	sourceStats := mock.NewEmptyStats(source)
+	sourceStats.AddProperty("tst", prop.ATKBase, 100.0)
+	sourceStats.AddProperty("tst", prop.DEFBase, 100.0)
+	sourceStats.AddProperty("tst", prop.HPBase, 100.0)
+	shield := &Instance{name: "SourceShield", hp: 100.0}
+	manager.targets[source] = append(manager.targets[source], shield)
+	attr.EXPECT().Stats(gomock.Eq(source)).Return(sourceStats).Times(5)
+
+	target := key.TargetID(2)
+	targetStats := mock.NewEmptyStats(target)
+	targetStats.AddProperty("tst", prop.HPBase, 100.0)
+	attr.EXPECT().Stats(gomock.Eq(target)).Return(targetStats).Times(5)
+
+	type shieldConfig struct {
+		ID   key.Shield
+		Info info.Shield
+	}
+
+	shieldConfigs := []shieldConfig{
+		{
+			ID: key.Shield("ShieldBySourceATK"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_ATK: 0.5},
+			},
+		},
+		{
+			ID: key.Shield("ShieldBySourceDEF"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_DEF: 0.5},
+			},
+		},
+		{
+			ID: key.Shield("ShieldBySourceHP"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_MAX_HP: 0.5},
+			},
+		},
+		{
+			ID: key.Shield("ShieldByTargetHP"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_TARGET_MAX_HP: 0.5},
+			},
+		},
+		{
+			ID: key.Shield("ShieldBySourceShield"),
+			Info: info.Shield{
+				Source:     source,
+				Target:     target,
+				BaseShield: info.ShieldMap{model.ShieldFormula_SHIELD_BY_SHIELDER_TOTAL_SHIELD: 0.5},
+			},
+		},
+	}
+
+	manager.event.ShieldAdded.Subscribe(func(event event.ShieldAdded) {
+		assert.Equal(t, 50.0, event.ShieldHealth)
+	})
```

Comment:
nit: should have different values for each case to make sure the wires are not crossed (right now could have `BY_TARGET_MAX_HP` using `ATK` and this test wouldn't catch it). Ideally you want 1 unit test only testing 1 thing so when it fails it is clear what is broken